plutus-core/src/Language/PlutusCore/Universe/Core.hs

```diff
@@ -104,17 +101,28 @@ knownUniOf _ = knownUni
 someValue :: forall a uni. uni `Includes` a => a -> Some (ValueOf uni)
 someValue = Some . ValueOf knownUni
 
--- | A universe is 'Closed', if it's known how to constrain every type from the universe
--- (the universe doesn't have to be finite for that).
+-- | A universe is 'Closed', if it's known how to constrain every type from the universe.
+-- The universe doesn't have to be finite and providing support for infinite universes is the
+-- reason why we encode a type as a sequence of integer tags as opposed to a single integer tag.
+-- For example, given
+--
+-- >   data U a where
+-- >       UList :: !(U a) -> U [a]
+-- >       UInt  :: U Int
+--
+-- @UList (UList UInt)@ can be encoded to @[0,0,1]@ where @0@ and @1@ are the integer tags of the
+-- @UList@ and @UInt@ constructors, respectively.
 class Closed uni where
     -- | A constrant for \"@constr a@ holds for any @a@ from @uni@\".
     type Everywhere uni (constr :: GHC.Type -> Constraint) :: Constraint
 
-    -- | Get the 'Int' tag of a type from @uni@. The opposite of 'uniAt'.
-    tagOf :: uni a -> Int
+    -- | Encode a type as a sequence of 'Int' tags.
+    -- The opposite of 'decodeUni'.
+    encodeUni :: uni a -> [Int]
```

Comment:
Perhaps `NonEmpty Int`?

Reply:
Then `NonEmpty Natural`. I don't feel like we need any safety here and `[]` has way better ergonomics than `NonEmpty`.